src/anderson_darling.jl

```diff
@@ -0,0 +1,172 @@
+# Anderson-Darling test
+
+export OneSampleADTest, KSampleADTest
+
+abstract ADTest <: HypothesisTest
+
+## ONE SAMPLE AD-TEST
+### http://www.itl.nist.gov/div898/handbook/eda/section3/eda35e.htm
+
+function adstats{T<:Real}(x::AbstractVector{T}, d::UnivariateDistribution)
+    n = length(x)
+    y = sort(x)
+    μ = mean(y)
+    σ = std(y)
+    z = cdf(d, (y-μ)/σ)
+    i = 1:n
+    S = ((2*i-1.0)/n) .* (log(z[i])+log(1-z[n+1-i]))
+    S[isinf(S)] = 0. # remove infinity
+    A² = -n-sum(S)
+    (n, μ, σ, A²)
+end
+
+immutable OneSampleADTest <: ADTest
+    n::Int      # number of observations
+    μ::Float64  # sample mean
+    σ::Float64  # sample std
+    A²::Float64 # Anderson-Darling test statistic
+end
+
+function OneSampleADTest{T<:Real}(x::AbstractVector{T}, d::UnivariateDistribution)
+    OneSampleADTest(adstats(x, d)...)
+end
+
+testname(::OneSampleADTest) = "One sample Anderson-Darling test"
+
+function show_params(io::IO, x::OneSampleADTest, ident="")
+    println(io, ident, "number of observations:   $(x.n)")
+    println(io, ident, "sample mean:              $(x.μ)")
+    println(io, ident, "sample SD:                $(x.σ)")
+    println(io, ident, "A² statistic:             $(x.A²)")
+end
+
+### Ralph B. D'Agostino, Goodness-of-Fit-Techniques, CRC Press, Jun 2, 1986
+### https://books.google.com/books?id=1BSEaGVBj5QC&pg=PA97, p.127
+function pvalue(x::OneSampleADTest)
+    z = x.A²*(1.0 + .75/x.n + 2.25/x.n/x.n)
+
+    # q = [.05, .10, .15, .20, .25, .30, .35, .40, .45, .50, .55, .60, .65, .70, .75, .80, .85, .90, .95, .975, .99, .995]
+    # b0 = [-.512, -.552, -.608, -.643, -.707, -.735, -.772, -.770, -.778, -.779, -.803, -.818, -.818, -.801, -.800, -.756, -.749, -.750, -.795, -.881, -1.013, -1.063]
+    # b1 =[2.10, 1.25, 1.07, .93, 1.03, 1.02, 1.04, .90, .80, .67, .70, .58, .42, .12, -.09, -.39, -.59, -.8, -.89, -.94, -.93, -1.34]
+    # a₊ = [.1674, .1938, .2147, .2333, .2509, .2681, .2853, .3030, .3213, .3405, .3612, .3836, .4085, .4367, .4695, .5091, .5597, .6305, .7514, .8728, 1.0348, 1.1578]
+    # a = a₊.*(1.+ b0/x.n +b1/x.n/x.n)
+    # qi = searchsortedlast(a, x.A²)
+    # Z(i) = a₊[i]*(1+b0[i]/x.n +b1[i]/x.n/x.n)
+
+    Z = [.200, .340, .600]
```

Comment:
Maybe avoid the array construction and extraction of values here and just substitute in the values.